Description: Add missing __acl_apply_mask_to_mode.c and fix remaining references
  The CVE-2026-54369.patch renamed __apply_mask_to_mode to
  __acl_apply_mask_to_mode in libacl.h and perm_copy_fd.c and updated
  the file reference in libacl/Makemodule.am, but:
  1. Never created the new source file __acl_apply_mask_to_mode.c
  2. Missed updating the call site in perm_copy_file.c:192
  This patch fixes both issues: creates the new source file, removes
  the old file, and updates the remaining call site.
Forwarded: not-needed
Last-Update: 2026-07-08
---

diff --git a/libacl/__apply_mask_to_mode.c b/libacl/__apply_mask_to_mode.c
deleted file mode 100644
--- a/libacl/__apply_mask_to_mode.c
+++ /dev/null
@@ -1,67 +0,0 @@
-/*
-  File: __apply_mask_to_mode.c
-  (Linux Access Control List Management)
-
-  Copyright (C) 1999-2002
-  Andreas Gruenbacher, <andreas.gruenbacher@gmail.com>
-
-  This library is free software; you can redistribute it and/or
-  modify it under the terms of the GNU Lesser General Public
-  License as published by the Free Software Foundation; either
-  version 2.1 of the License, or (at your option) any later version.
-
-  This library is distributed in the hope that it will be useful,
-  but WITHOUT ANY WARRANTY; without even the implied warranty of
-  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-  Lesser General Public License for more details.
-
-  You should have received a copy of the GNU Lesser General Public
-  License along with this library; if not, see <https://www.gnu.org/licenses/>.
-*/
-
-#include "config.h"
-#include <sys/stat.h>
-#include "libacl.h"
-
-#if defined(HAVE_ACL_ENTRIES) && \
-    defined (HAVE_ACL_GET_ENTRY) && defined(HAVE_ACL_GET_TAG_TYPE) && \
-    defined (HAVE_ACL_GET_PERMSET) && defined(HAVE_ACL_GET_PERM)
-int
-__apply_mask_to_mode(mode_t *mode, acl_t acl)
-{
-	acl_entry_t entry;
-	int entry_id=ACL_FIRST_ENTRY;
-
-	/* A minimal ACL which has three entries has no mask entry; the
-	   group file mode permission bits are exact. */
-	if (acl_entries(acl) == 3)
-		return 0;
-
-	while (acl_get_entry(acl, entry_id, &entry) == 1) {
-		acl_tag_t tag_type;
-
-		acl_get_tag_type(entry, &tag_type);
-		if (tag_type == ACL_MASK) {
-			acl_permset_t permset;
-
-			acl_get_permset(entry, &permset);
-			if (acl_get_perm(permset, ACL_READ) != 1)
-				*mode &= ~S_IRGRP;
-			if (acl_get_perm(permset, ACL_WRITE) != 1)
-				*mode &= ~S_IWGRP;
-			if (acl_get_perm(permset, ACL_EXECUTE) != 1)
-				*mode &= ~S_IXGRP;
-
-			return 0;
-		}
-		entry_id = ACL_NEXT_ENTRY;
-	}
-
-	/* This is unexpected; if the ACL didn't include a mask entry
-	   we should have exited before the loop! */
-	*mode &= ~S_IRWXG;
-	return 1;
-}
-#endif
-
-
diff --git a/libacl/perm_copy_file.c b/libacl/perm_copy_file.c
--- a/libacl/perm_copy_file.c
+++ b/libacl/perm_copy_file.c
@@ -189,7 +189,7 @@ perm_copy_file (const char *src_path, const char *dst_path,
 
 	if (acl_set_file (dst_path, ACL_TYPE_ACCESS, acl) != 0) {
 		int saved_errno = errno;
-		__apply_mask_to_mode(&st.st_mode, acl);
+		__acl_apply_mask_to_mode(&st.st_mode, acl);
 		ret = chmod (dst_path, st.st_mode);
 		if ((errno != ENOSYS && errno != ENOTSUP) ||
 		    acl_entries (acl) != 3) {

diff --git a/libacl/__acl_apply_mask_to_mode.c b/libacl/__acl_apply_mask_to_mode.c
new file mode 100644
--- /dev/null
+++ b/libacl/__acl_apply_mask_to_mode.c
@@ -0,0 +1,67 @@
+/*
+  File: __acl_apply_mask_to_mode.c
+  (Linux Access Control List Management)
+
+  Copyright (C) 1999-2002
+  Andreas Gruenbacher, <andreas.gruenbacher@gmail.com>
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "config.h"
+#include <sys/stat.h>
+#include "libacl.h"
+
+#if defined(HAVE_ACL_ENTRIES) && \
+    defined (HAVE_ACL_GET_ENTRY) && defined(HAVE_ACL_GET_TAG_TYPE) && \
+    defined (HAVE_ACL_GET_PERMSET) && defined(HAVE_ACL_GET_PERM)
+int
+__acl_apply_mask_to_mode(mode_t *mode, acl_t acl)
+{
+	acl_entry_t entry;
+	int entry_id=ACL_FIRST_ENTRY;
+
+	/* A minimal ACL which has three entries has no mask entry; the
+	   group file mode permission bits are exact. */
+	if (acl_entries(acl) == 3)
+		return 0;
+
+	while (acl_get_entry(acl, entry_id, &entry) == 1) {
+		acl_tag_t tag_type;
+
+		acl_get_tag_type(entry, &tag_type);
+		if (tag_type == ACL_MASK) {
+			acl_permset_t permset;
+
+			acl_get_permset(entry, &permset);
+			if (acl_get_perm(permset, ACL_READ) != 1)
+				*mode &= ~S_IRGRP;
+			if (acl_get_perm(permset, ACL_WRITE) != 1)
+				*mode &= ~S_IWGRP;
+			if (acl_get_perm(permset, ACL_EXECUTE) != 1)
+				*mode &= ~S_IXGRP;
+
+			return 0;
+		}
+		entry_id = ACL_NEXT_ENTRY;
+	}
+
+	/* This is unexpected; if the ACL didn't include a mask entry
+	   we should have exited before the loop! */
+	*mode &= ~S_IRWXG;
+	return 1;
+}
+#endif
+
+
